Add Korean translation

keispace opened the pull request:
about #262
This is a Korean translation draft file(ko.po)  that reflects the commitment of f3446a91(25 jan 2023)

I'm sorry that it seems to be a similar PR(#276) duplication, but there is a difference in the scope of the work, so I'm uploading the PR. 
I added it by referring to the @jiyongp comment in #276. Thank you. 
However, as a developer in Korea, there were some opinions, so I couldn't reflect them all. 
Please let me know if anyone have any additional comments on the following. Or anything else.


- Copying and Cloning: I don't think we need to use the original language as it is, as it can be divided into 복사 and 복제 rather than copying the word itself. 
- Shared and Unique Borrows: Translating Unique into '단독' would confuse with single.
- Designing a Library: It's an actual book library system, so I kept it as a '도서관' 
- Niche Optimization: '소소한' doesn't convey the meaning of niche, and in Korea, and Korean developers' blogs use it as 'niche Optimization'. so I'm maintaining it for now.
- Paths: I don't think it's a separate type of Paths, it's an absolute/relative path, so I kept it as a path. It will be applied after checking again.
- Trait: I understood it as a big characteristic of Rust, so I got a tuning fork in the original language. The word "특성" comes in general meanings like attributes and properties.
- DEFERENCING Raw Pointers: I agree, but I'm writing it in '역참조(따라가기)' because I think it's right to explain it in terms that are understandable to existing developers. 
- Mutable Static Variables: '가변 정적 변수' / '정적 가변 변수' all have the same meaning, and I thought write '가변' in sentence form, I think '정적' should be the same. so I kept it.

jiyongp commented on po/ko.po:
```diff
@@ -146,19 +146,19 @@ msgstr "타입 추론"
 #: src/SUMMARY.md:43
 msgid "static & const"
-msgstr ""
+msgstr "정적변수(static)과 상수(const)"
```
nit: s/과/와/

jiyongp commented on po/ko.po:
```diff
@@ -146,19 +146,19 @@ msgstr "타입 추론"
 #: src/SUMMARY.md:44
 msgid "Scopes and Shadowing"
-msgstr "스코프(범위)와 쉐도잉"
+msgstr "스코프 범위(Scopes)와 쉐도잉(Shadowing)"
```
drop 스코프

jiyongp commented on po/ko.po:
```diff
@@ -146,19 +146,19 @@ msgstr "타입 추론"
 #: src/SUMMARY.md:46
 msgid "Stack vs Heap"
-msgstr ""
+msgstr "Stack과 Heap"
```
스택(Stack)과 힙(Heap)

jiyongp commented on po/ko.po:
```diff
@@ -190,19 +190,19 @@ msgstr "소유권"
 #: src/SUMMARY.md:57
 msgid "Moves in Function Calls"
-msgstr "함수호출에서 이동"
+msgstr "함수호출에서의 이동(Move)"
```
함수호출 or 함수 호출? We should be consistent.

keispace replied:
sorry; 함수 호출 is correct 🥲

jiyongp commented on po/ko.po:
```diff
@@ -546,7 +546,7 @@ msgstr "안전하지 않은 러스트"
 #: src/SUMMARY.md:164
 msgid "Dereferencing Raw Pointers"
-msgstr "원시포인트 역참조"
+msgstr "원시포인트 역참조(따라가기)"
```
s/포인트/포인터/

jiyongp commented on po/ko.po:
```diff
@@ -706,27 +706,27 @@ msgstr "C에서 러스트 호출"
 #: src/SUMMARY.md:217
 msgid "With C++"
-msgstr ""
+msgstr "C++과의 상호운용성"
```
C++와의 상호운용성.

jiyongp commented on po/ko.po:
```diff
@@ -4381,7 +4382,7 @@ msgstr ""
 #: src/ownership/moves-function-calls.md:1
 msgid "# Moves in Function Calls"
-msgstr "# 함수호출에서의 이동"
+msgstr "# 함수호출에서의 이동(Move)"
```
Let's be consistent in spacing

jiyongp commented on po/ko.po:
```diff
@@ -10082,7 +10083,7 @@ msgid ""
 "* Implement `unsafe` traits."
 msgstr ""
-"* 원시포인트 역참조\n"
+"* 원시포인트 역참조(따라가기)\n"
```
포인터

jiyongp commented on po/ko.po:
```diff
@@ -10101,11 +10102,11 @@ msgstr ""
 #: src/unsafe/raw-pointers.md:1
 msgid "# Dereferencing Raw Pointers"
-msgstr "# 원시포인트 역참조"
+msgstr "# 원시포인트 역참조(따라가기)"
```
포인터

jiyongp commented on po/ko.po:
```diff
@@ -10101,11 +10102,11 @@ msgstr ""
 #: src/unsafe/raw-pointers.md:3
 msgid "Creating pointers is safe, but dereferencing them requires `unsafe`:"
-msgstr "포인트 생성은 안전합니다. 하지만 역참조할 경우 `unsafe`가 필요합니다:"
+msgstr "포인트 생성은 안전합니다. 하지만 역참조(따라가기)할 경우 `unsafe`가 필요합니다:"
```
포인터

jiyongp commented:
> * Copying and Cloning: I don't think we need to use the original language as it is, as it can be divided into 복사 and 복제 rather than copying the word itself.

Ack.

> * Shared and Unique Borrows: Translating Unique into '단독' would confuse with single.
> * Designing a Library: It's an actual book library system, so I kept it as a '도서관'

Oh I missed that. Thanks for correcting.

> * Niche Optimization: '소소한' doesn't convey the meaning of niche, and in Korea, and Korean developers' blogs use it as 'niche Optimization'. so I'm maintaining it for now.

Ok.

> * Paths: I don't think it's a separate type of Paths, it's an absolute/relative path, so I kept it as a path. It will be applied after checking again.

You are right.

> * Trait: I understood it as a big characteristic of Rust, so I got a tuning fork in the original language. The word "특성" comes in general meanings like attributes and properties.

I am fine with 트레이트(Trait). Note that we should drop s from Traits.

> * DEFERENCING Raw Pointers: I agree, but I'm writing it in '역참조(따라가기)' because I think it's right to explain it in terms that are understandable to existing developers.

Ack. We can't fix the existing habit. :(

> * Mutable Static Variables: '가변 정적 변수' / '정적 가변 변수' all have the same meaning, and I thought write '가변' in sentence form, I think '정적' should be the same. so I kept it.

jiyongp commented on po/ko.po:
```diff
@@ -2121,11 +2120,12 @@ msgid ""
 "* [`std::mem::forget`]을 사용하여 컴파일러가 값에 대해 \"잊도록\" 만들 수 있"
 "습니다(소멸자가 실행되지 않음을 의미합니다).\n"
-"* `Rc` 또는 `Arc`를 사용하여 실수로 [reference cycle]을 생성할 수도 있습니다."
+"* `Rc` 또는 `Arc`를 사용하여 실수로 [reference cycle]을 생성할 수도 있습니다.\n"
```
순환참조

jiyongp commented on po/ko.po:
```diff
@@ -2726,11 +2724,18 @@ msgid ""
-"지 '살아있어야' 슬라이스를 가져올 수 있습니다."
+"* `s`는 `i32` 슬라이스에 대한 참조입니다. `s`의 타입(`&[i32]`)은 더이상 배열의"
+" 길이에 대해 다루지 않게 되며 이를 통해 다양한 크기의 슬라이스에 대한 계산을"
+" 수행할 수 있습니다.\n"
```
I think we need to be creative when translating this part.

s의 타입(...)에 배열의 크기가 빠져 있음을 주목하세요. 즉, 슬라이스는 임의의 길이를 가지는 데이터를 다룰 수 있게 해 줍니다.

jiyongp commented on po/ko.po:
```diff
@@ -2726,11 +2724,18 @@ msgid ""
+" (범위 내에서) '살아있어야' 슬라이스를 가져올 수 있습니다.\n"
+"    \n"
+"* `a[3]`로 수정하라는 질문은 흥미로운 토론을 불러 일으킬 수 있지만,"
+" 결론은 메모리 안전을 위해 슬라이스를 만든 후에는 `a`를 수정 할 수 없지만"
```
`a[3]`의 값을 바꿀 수 있냐는 질문은 좋은 질문입니다. 여기에 대한 답은 a와 s를 통해 데이터를 읽을 수는 있지만 수정할 수는 없으며, 이는 메모리 안전을 위해서라는 것입니다. 왜 그런지에 대한 좀더 구체적인 답은 빌림 검사 부분에서 자세히 설명합니다.

jiyongp commented on po/ko.po:
```diff
@@ -2776,19 +2781,17 @@ msgstr ""
-"니다.\n"
-"문자열 리터럴(`”Hello”`)은 프로그램 바이너리에 저장됩니다. "
+"* `&str`은 메모리 블록에 저장된 UTF-8인코딩 문자열 데이터의 불변 참조 슬라이스"
+" 입니다.\n"
```
`&str`은 문자열 슬라이스에 대한 (불변) 참조입니다. 러스트에서 문자열은 UTF-8로 인코딩 된 데이터를 의미합니다.

jiyongp commented on po/ko.po:
```diff
@@ -2798,9 +2801,11 @@ msgid ""
-"* `String::from`는 문자 리터럴(`&str`)로 부터 문자열을 생성합니다. `String::"
-"new`는 새로운 빈 문자열을 생성합니다. 여기에 `to_string`메서드를 이용해 문자"
-"열을 추가할 수 있습니다."
+"* 러스트의 `String` 타입은 바이트 벡터 래퍼입니다. `Vec<T>`와 같이 소유됩니다.\n"
```
러스트의 `String`타입은 실제로는 문자열을 이루는 바이트에 대한 백터(`Vec<u8>`)입니다. `Vec<T>`가 `T`를 소유하고 있듯이, `String`이 가리키고 있는 문자열은  `String`의 소유입니다.

jiyongp commented on po/ko.po:
```diff
@@ -2934,10 +2938,16 @@ msgid ""
-"서는 반대입니다.)* 함수 본문의 마지막 표현식은 반환 값이 됩니다. 간단히, 식 "
-"끝에 있는 `;`를 생략하면 됩니다.* 반환값이 없는 함수의 경우, `유닛 타입 ()`"
-"을 반환합니다. `-> ()`가 생략된 경우 컴파일러는  이를 추론합니다."
+"* 매개변수 선언 뒤에는 타입 표시합니다. 다음에는 반환 타입입니다.(일부 언어에"
```
매개변수를 선언할 때에는 이름을 먼저 쓰고, 타입을 나중에 씁니다. 이름과 타입은 `:` 로 구분합니다. 이는 일부 언어(예를 들어 C)와 반대임에 유의하세요. 마찬가지로, 리턴 타입도 함수의 시작이 아닌 가장 뒷부분에 선언합니다.

jiyongp commented on po/ko.po:
```diff
@@ -10938,6 +10953,17 @@ msgid ""
+"<details>\n"
+"    \n"
+"* 러스트의 `Mutex`는 보호된 데이터라는 하나의 요소만으로 이루어진 컬렉션처럼"
+" 동작합니다.\n"
```
러스트의 `Mutex`는 오직 하나의 데이터만 담을 수 있는 컬렉션으로 볼 수도 있습니다. 다른 컬렉션과 다른 점은, 그 데이터가 동시성 문제로부터 보호된다는 점입니다.

jiyongp commented on po/ko.po:
```diff
@@ -10938,6 +10953,17 @@ msgid ""
+"    \n"
+"* 러스트의 `Mutex`는 보호된 데이터라는 하나의 요소만으로 이루어진 컬렉션처럼"
+" 동작합니다.\n"
+"  * 보호된 데이터에 접근하기 전 `Mutex`를 획득하는 것을 잊지 않아야 합니다.\n"
```
Mutex는 mutex를 획득하지 않고서는 그 데이터에 접근하는 것이 원천적으로 불가능 하도록 디자인 되어 있습니다.

jiyongp commented on po/ko.po:
```diff
@@ -10938,6 +10953,17 @@ msgid ""
+"* 러스트의 `Mutex`는 보호된 데이터라는 하나의 요소만으로 이루어진 컬렉션처럼"
+" 동작합니다.\n"
+"  * 보호된 데이터에 접근하기 전 `Mutex`를 획득하는 것을 잊지 않아야 합니다.\n"
+"* 읽기-쓰기에 대한 잠금은 `RwLock`가 있습니다.\n"
```
I think this needs some elaboration.

`Mutex`는 읽기 잠금과 쓰기 잠금을 구별하지 않습니다. 때문에 여러 스레드에서 데이터를 "읽기"만 하려고 해도 이를 동시에 지원할 수 없습니다. 한 순간에 한 스레드에서의 읽기만 허용이 됩니다. "읽기"에 대한 동시성을 지원하면서 "쓰기"에 대해서 안전하고 싶다면 `RwLock`을 사용할 수 있습니다.

jiyongp commented on po/ko.po:
```diff
@@ -10938,6 +10953,17 @@ msgid ""
+"  * 보호된 데이터에 접근하기 전 `Mutex`를 획득하는 것을 잊지 않아야 합니다.\n"
+"* 읽기-쓰기에 대한 잠금은 `RwLock`가 있습니다.\n"
+"* 어째서 `lock()`이 `Result`를 반환합니까?\n"
+"  * `Mutex`를 고정한 스레드가 패닉 상태가 되면, `Mutex`는 보호한 데이터가"
```
`Mutex`를 획득한 스레드에서 패닉이 발생하면, 데이터가 올바르지 않은 상태에 머물러 있을 수 있습니다. 이런 상황을  `Mutex`가 중독(poisoned)되었다고 하며, 중독된 뮤텍스에서 ...

jiyongp commented on po/ko.po:
```diff
@@ -13296,6 +13350,8 @@ msgid ""
 "the same length, so such variable could contain an invalid matrix."
 msgstr ""
+"또한 타임 자체의 슬라이스 길이가 동일하지 않으므로 이러한 변수에 잘못된 행렬이"
+" 포함될 수 있습니다."
```
또한, 슬라이스 타입은 길이를 포함하지 않기 때문에 한 단계 아래의 슬라이스들이 같은 길이(역주: 3)임을 보장할 수 없습니다. 때문에 슬라이스 타입의 변수에는 잘못된 행렬(역주: 각 열이나 행의 길이가 다른)이 전달될 수 있습니다.

jiyongp reviewed (commented):
I'd like to merge this series sooner than later. So please tell me when you are ready. There might be more places to fix (I haven't read all the translations yet), but I think those could be fixed later.

@mgeisler do you agree?

keispace commented:
> I'd like to merge this series sooner than later. So please tell me when you are ready. There might be more places to fix (I haven't read all the translations yet), but I think those could be fixed later.

Yes, there are quite a few ambiguous meanings (because it's only been a few weeks since I studied Rust), but the PR is now up-to-date with the latest version of the upstream.
I think it would be good to continue to get reviews from many people and improve it.

mgeisler commented:
> I'd like to merge this series sooner than later. So please tell me when you are ready. There might be more places to fix (I haven't read all the translations yet), but I think those could be fixed later.
> 
> @mgeisler do you agree?

I'm happy to merge this! Thanks so much @keispace for the work and thank you @jiyongp for proof-reading it!'

We can merge this now, and then we can enable the publication of the translation on Monday (by editing [`publish.yml`](https://github.com/google/comprehensive-rust/blob/main/.github/workflows/publish.yml#L21)). The idea is that we can de-couple the development of a translation (which can take a long time) from the publication of the translation.

mgeisler commented:
@keispace, can you add a line with

```
- ko
```

to the [`build.yml` file](https://github.com/google/comprehensive-rust/blob/main/.github/workflows/build.yml#L86)?

Based on @jiyongp's comments, I think you should also go ahead and update `publish.yml` so that we can publish this right away next week :tada:

keispace commented:
> @keispace, can you add a line with
> 
> ```
> - ko
> ```
> 
> to the [`build.yml` file](https://github.com/google/comprehensive-rust/blob/main/.github/workflows/build.yml#L86)?
> 
> Based on @jiyongp's comments, I think you should also go ahead and update `publish.yml` so that we can publish this right away next week 🎉

@mgeisler 
I'm not familiar with the Git workflow and I'm careful to modify it other than the translation file, so I just modified the po file for now. I've added the part you mentioned. 
And I want to add it to the 'LANGUAGES' in the `publish.yml` file, but I left it for now because there was an error in writing the list.

jiyongp commented:
@keispace, please fix the errors from `Test/translations (ko)`. There are multiple places where the newline character is missing.

Or you may follow the instructions at https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork (which is basically selecting the checkbox "Allow edits from maintainers") so that I can make the fix by myself.

jiyongp commented on po/ko.po:
```diff
@@ -0,0 +1,14439 @@
+"about Rust and hope to:"
+msgstr ""
+"강의는 당신이 러스트에 대해서 아무것도 모른다고 가정하고 아래의 목표를가지고 "
+"있습니다. \n"
```
```suggestion
"있습니다."
```

jiyongp commented on po/ko.po:
```diff
@@ -0,0 +1,14439 @@
+
+#: src/welcome.md:14
+msgid "On Day 4, we will cover Android-specific things such as:"
+msgstr ""
+"4일차 강의에 우리는 아래와 같은 안드로이드 특화된 내용들도 설명합니다.\n"
```
```suggestion
msgstr "4일차 강의에 우리는 아래와 같은 안드로이드 특화된 내용들도 설명합니다."
```

jiyongp commented on po/ko.po:
````diff
@@ -0,0 +1,14439 @@
+"   checkout. This will ensure that the Android build system sees the\n"
+"   `Android.bp` files in `src/android/`."
+msgstr ""
+"6. 4일차에 안드로이드 부분을 건너띄지 않으려면 [AOSP checkout][1]이 필요합니"
+"다. 동일한 컴퓨터에 있는 [과정 저장소][2]를 체크아웃하고 `src/android/` 디렉"
+"토리를 AOSP 체크아웃의 루트로 이동합니다. 이렇게 하면 Android 빌드 시스템에"
+"서 `src/android/`의 `Android.bp` 파일을 볼 수 있습니다.\n"
+"  * _역주: 안드로이드쪽은 지식이 짧아 정확히 모르겠네요 원문 첨부합니다._\n"
+"    ```\n"
+"      6. If you don't skip the Android specific parts on Day 4, you will "
+"need an         [AOSP\n"
+"        checkout][1]. Make a checkout of the [course repository][2] on the "
+"same\n"
+"        machine and move the `src/android/` directory into the root of your "
+"AOSP\n"
+"        checkout. This will ensure that the Android build system sees the\n"
+"        `Android.bp` files in `src/android/`. \n"
+"    ```\n"
````
```suggestion
"6. 4일차에 있을 안드로이드 부분을 건너뛰지 않는다면, [AOSP 코드를][1] 체크아웃 해야\n"
"합니다. 그런 동일한 그 컴퓨터에 [과정 저장소][2]를 체크아웃하고 과정 저장소에 있는 \n"
"`src/android/` 디렉터리를 AOSP 소스코드의 루트로 이동합니다. 이렇게 하면 안드로이드 \n"
"빌드 시스템에서 `src/android/`의 `Android.bp` 파일을 인식할 수 있습니다."
```

jiyongp commented on po/ko.po:
```diff
@@ -0,0 +1,14439 @@
+#: src/error-handling/converting-error-types.md:55
+#: src/concurrency/threads.md:30
+msgid "Key points:"
+msgstr "키 포인트: \n"
```
```suggestion
msgstr "키 포인트:"
```

jiyongp commented:
I've merged #300 which is an amendment of this PR. Let's close this here. Thanks @keispace for your contribution.
